core/src/main/java/com/sequenceiq/cloudbreak/service/StackUpdater.java

```diff
@@ -112,14 +93,55 @@ public void updateVariant(Long resourceId, String variant) {
         }
     }
 
-    private void saveDeprecatedClusterStatus(String statusReason, Stack stack, Status newStatus) {
-        Cluster cluster = stack.getCluster();
+    private void doUpdateStackStatus(Long stackId, Status newStatus, DetailedStackStatus newDetailedStatus, String statusReason) {
+        try {
+            StackBase stackBase = transactionService.required(() -> {
+                StackBase stack = stackService.getStackBaseById(stackId);
+                StackStatus<StackBase> actualStackStatus = stack.getStackStatus();
+                LOGGER.info("Update stack status from: {}/{} to: {}/{} stack: {} reason: {}", actualStackStatus.getStatus(),
+                        actualStackStatus.getDetailedStackStatus(),
+                        newStatus, newDetailedStatus, stackId, statusReason);
+                if (Objects.equals(actualStackStatus.getStatus(), newStatus)) {
+                    LOGGER.debug("New status is the same as previous status {}/{}, skip status update.",
+                            actualStackStatus.getStatus(), actualStackStatus.getDetailedStackStatus());
+                    return null;
+                } else if (Status.DELETE_COMPLETED.equals(actualStackStatus.getStatus())) {
+                    LOGGER.info("Stack is in DELETE_COMPLETED status, cannot update status.");
+                    return null;
+                } else {
+                    StackStatus<StackBase> stackStatus = new StackStatus<>(stack, newStatus, statusReason, newDetailedStatus);
+                    stack.setStackStatus(stackStatus);
+                    StackBase savedStack = stackService.update(stack);
+                    saveDeprecatedClusterStatus(newStatus, statusReason, stack.getCluster());
+                    if (newStatus.isRemovableStatus()) {
+                        InMemoryStateStore.deleteStack(stackId);
+                        if (stack.getCluster() != null) {
+                            InMemoryStateStore.deleteCluster(stack.getCluster().getId());
+                        }
+                    } else {
+                        InMemoryStateStore.putStack(stackId, statusToPollGroupConverter.convert(newStatus));
+                        if (stack.getCluster() != null) {
+                            InMemoryStateStore.putCluster(stack.getCluster().getId(), statusToPollGroupConverter.convert(newStatus));
+                        }
+                    }
+                    return savedStack;
+                }
+            });
+            if (stackBase != null) {
+                usageLoggingUtil.logClusterStatusChangeUsageEvent(stackBase.getStackStatus().getStatus(), newStatus, stackBase);
```

Comment:
At this point the `stackBase.getStackStatus().getStatus()` will always be equals to `newStatus`, or am I wrong?